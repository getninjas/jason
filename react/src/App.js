import React, { Component } from 'react';
import Form from './Form'
import request from './services/request'

class App extends Component {
  state = {
    data: '',
    type: ''
  }

  componentDidMount() {
    request().then(data => this.setState({ data: data._embedded, type: data.type }));
  }

  render() {
    return (!!this.state.data ? <Form data={this.state.data} type={this.state.type} />  : '' );
  }
}

export default App;
