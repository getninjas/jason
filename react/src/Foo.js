import React, { Component } from 'react';

class Foo extends Component {
  render() {
    return (
      <div>
        Teste
        <h1>Hello</h1>
      </div>
    );
  }
}

export default Foo;
