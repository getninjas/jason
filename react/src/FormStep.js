import React, { Component } from 'react';

class FormStep extends Component {

  state = {
    formData: this.props.valueInput
  }

  render() {
    return (
      <div id={this.props.index}>
        <h2>Step {this.props.index}</h2>

        {this.props.step.map((field, i) => {
          return (
            <div key={i}>
              <label htmlFor={field.name}>{field.label}</label><br />
              <input
                type="text"
                placeholder={field.mask}
                required={field.required}
                value={this.state.formData[field.name]}
                name={field.name}
                onChange={this.props.dataHandler} />
            </div>
          )
        })}

        {this.props.index !== 0 && this.props.stepBy && (
          <button type="button" onClick={this.props.onBackClick}>Back</button>
        )}

        <button>Next</button>
      </div>
    );
  }
}

export default FormStep;
