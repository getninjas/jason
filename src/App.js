import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import Form from './components/Form';

const propTypes = {
  data: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  action: PropTypes.string.isRequired,
  onZipcodeFetchSuccess: PropTypes.func,
  onZipcodeFetchError: PropTypes.func,
  onSubmit: PropTypes.func,
  onSubmitSuccess: PropTypes.func,
  onSubmitError: PropTypes.func,
  onStepChange: PropTypes.func,
};

const defaultProps = {
  onReady() {},
  onZipcodeFetchSuccess() {},
  onZipcodeFetchError() {},
  onSubmit() {},
  onSubmitSuccess() {},
  onSubmitError() {},
  onStepChange() {},
};

export default class App extends Component {
  constructor() {
    super();
    this.ref = createRef();
  }

  post() {
    this.ref.current.formSubmit();
  }

  render() {
    return <Form {...this.props} data={this.props.data} />;
  }
}

App.propTypes = propTypes;
App.defaultProps = defaultProps;
