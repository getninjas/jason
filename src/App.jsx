import React from 'react';
import PropTypes from 'prop-types';
import Select from './components/Select';
import ParserFields from './lib/ParserFields';

export default class App extends React.Component {
  constructor() {
    super();

    this.state = {
      data: [{
        values: []
      }]
    }
  }

  componentDidMount() {
    const data = new ParserFields(this.props.jsonFields).init();

    this.setState({ data });
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Welcome to Jason!</h1>
        </header>

        <Select id={'teste-select'} values={this.state.data[0].values} name={'my-custom-select'}/>
      </div>
    );
  }
}

App.propTypes = {
  jsonFields: PropTypes.object.isRequired
}
