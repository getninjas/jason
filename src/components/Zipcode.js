import axios from 'axios';
import IMask from 'imask';
import PropTypes from 'prop-types';
import React, { Component, createRef, Fragment } from 'react';
import { AppContext } from '../AppContext';
import triggerNativeEvent from '../helpers/domEvent';
import { fillAddressState, getEmptyState, isUserTyping, isValidZipCodeInput } from '../helpers/zipcode';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  style: PropTypes.string,
  initialValue: PropTypes.string,
  value: PropTypes.any,
  zipcodeUrlService: PropTypes.string.isRequired,
  minLength: PropTypes.number,
  mask: PropTypes.string,
};

const defaultProps = {
  placeholder: '99999-999',
  required: true,
  title: 'Zipcode',
  type: 'zipcode',
  style: 'form__input',
  value: '',
  minLength: 9,
};

export default class Zipcode extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: this.props.initialValue || defaultProps.value,
      type_street: '',
      street: '',
      city: '',
      neighborhood: '',
      uf: '',
      fullAddress: '',
      zipcodeInvalid: false,
      fetchCompleted: false,
      zipcodeUrlService: this.props.zipcodeUrlService,
      fetching: false,
    };

    this.mounted = false;
    this.inputRef = createRef();
    this.onBlur = this.onBlur.bind(this);
    this.triggerEvent = triggerNativeEvent;
  }

  updateZipcode(zipcode, successCallback, errorCallback) {
    this.setState({ fetching: true, value: zipcode });
    this.getZipCode(zipcode, successCallback, errorCallback);
  }

  onChange(successCallback, errorCallback, evt) {
    const zipcode = evt.target.value;

    if (isUserTyping(zipcode.length, this.props.minLength)) {
      this.setState({ value: zipcode, fullAddress: '', fetchCompleted: false });
    } else if (isValidZipCodeInput(zipcode.length, this.props.minLength, this.state.fetchCompleted)) {
      this.props.onFieldBlur({ ...this.props, value: zipcode });

      this.updateZipcode(zipcode, successCallback, errorCallback);
    }
  }

  onBlur(successCallback, errorCallback, evt) {
    const zipcode = evt.target.value;
    const { fetchCompleted, zipcodeInvalid } = this.state;

    this.props.onFieldBlur({ ...this.props, value: this.state.zipcodeInvalid ? '' : zipcode });

    if (isValidZipCodeInput(zipcode.length, this.props.minLength, fetchCompleted) && !zipcodeInvalid) {
      this.updateZipcode(zipcode, successCallback, errorCallback);
    }
  }

  async getZipCode(zipcode, successCallback, errorCallback) {
    try {
      const url = this.state.zipcodeUrlService.replace(/@@zipcode@@/, zipcode.replace(/[^0-9]/g, ''));
      const response = await axios.get(url);

      this.onZipcodeSuccess(zipcode, response);

      successCallback(this.state);
    } catch (error) {
      this.onZipcodeError();

      errorCallback({ ...this.state, error });
    }
  }

  onZipcodeSuccess(zipcode, response) {
    let result = getEmptyState(this.state);

    result = fillAddressState(response.data, zipcode);
    result.fetching = false;
    result.zipcodeUrlService = this.props.zipcodeUrlService;
    result.zipcodeInvalid = false;

    this.setState(result);

    this.props.onFieldChange({ ...this.props, value: zipcode });
  }

  onZipcodeError() {
    const { value } = this.state;
    this.setState({ fullAddress: '', fetching: false, zipcodeInvalid: true });

    this.props.onFieldBlur({ ...this.props, fetchCompleted: false, value: this.state.zipcodeInvalid ? '' : value });
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  componentDidMount() {
    this.mounted = true;
    this.mask = new IMask(this.inputRef.current, { mask: this.props.mask });

    this.mask.on('complete', () => {
      this.props.onFieldChange({
        value: this.mask.value,
        id: this.props.id,
      });

      if (this.mounted) {
        this.setState({ value: this.mask.value, zipcodeInvalid: false });
      }
    });

    if (this.state.value.length) {
      this.triggerEvent(`#${this.props.id}`, 'blur');
    }
  }

  render() {
    const { id, name, required, placeholder, style } = this.props;
    const { street, city, neighborhood, uf, fullAddress, fetching } = this.state;
    const searchingZipcode = <span className='zipcode__loader' >Buscando CEP...</span>;
    const completeAddress = <span className='full-address'>{fullAddress}</span>;

    return (
      <AppContext.Consumer>
        { context => <Fragment>
          <a
            href='http://www.buscacep.correios.com.br'
            target='_blank'
            className='form__label-link'
            rel='noopener noreferrer'>
            NÃ£o lembra seu CEP?
          </a>
          <input
            id={id}
            name={name}
            className={style}
            type='tel'
            placeholder={placeholder}
            required={required}
            onChange={this.onChange.bind(this, context.onZipcodeFetchSuccess, context.onZipcodeFetchError)}
            onBlur={this.onBlur.bind(this, context.onZipcodeFetchSuccess, context.onZipcodeFetchError)}
            ref={this.inputRef} value={this.state.value}
          />
          { fetching ? searchingZipcode : completeAddress }
          <input id='street' name='street' type='hidden' value={street} />
          <input id='neighborhood' name='neighborhood' type='hidden' value={neighborhood} />
          <input id='city' name={'city'} type={'hidden'} value={city} />
          <input id='uf' name='uf' type='hidden' value={uf} />
        </Fragment>
        }
      </AppContext.Consumer>
    );
  }
}

Zipcode.propTypes = propTypes;
Zipcode.defaultProps = defaultProps;
