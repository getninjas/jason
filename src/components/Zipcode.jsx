import React, { Component, Fragment, createRef } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import IMask from 'imask';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  style: PropTypes.string,
  value: PropTypes.any,
  minLength: PropTypes.number,
};

const defaultProps = {
  placeholder: '99999-999',
  required: true,
  title: 'Zipcode',
  type: 'zipcode',
  style: 'form__input',
  value: '',
  minLength: 9,
};

export default class Zipcode extends Component {
  constructor() {
    super();

    this.state = {
      value: '',
      type_street: '',
      street: '',
      city: '',
      neighborhood: '',
      uf: '',
      fullAddress: '',
    }

    this.ref = createRef();
    this.onKeyUp = this.onKeyUp.bind(this);
  }

  onKeyUp(evt) {
    const zipcode = evt.target.value.replace(/[^0-9]/g, '');
    const key = Number(evt.key);

    this.props.onFieldChange({ ...this.props, value: evt.target.value });

    if (!isNaN(key)) {
      if (zipcode.length === 8) {
        this.getZipCode(zipcode);
      }
    }
  }

  getZipCode(zipcode) {
    let result = this.getEmptyState(this.state);

    axios.get(`http://www.getninjas-homolog.com.br/api/correios?q=${zipcode}`)
      .then((response) => {
        result = this.fillAddressState(response.data, zipcode);

        this.setState(result)
      })
      .catch(() => {
        result.value = zipcode;

        this.props.onFieldChange({ value: '', ...this.props});

        this.setState(result);
      });
  }

  getEmptyState(state) {
    const result = Object.keys(state).reduce((output, key) => {
      return Object.assign({}, output, { [key]: ''});
    }, {});

    return result;
  }

  fillAddressState(responseAddress, zipcode) {
    let result = Object.keys(responseAddress).reduce((output, key) => {
      return Object.assign({}, output, { [key]: responseAddress[key] });
    }, {});

    result.value = zipcode;
    result.fullAddress = this.getFullAddress(responseAddress);

    return result;
  }

  getFullAddress({street, neighborhood, city, uf}) {
    return `${street}, ${neighborhood} \n${city} - ${uf}`;
  }

  componentDidMount() {
    new IMask(this.ref.current, { mask: '00000-000' });

    this.setState({ value: this.props.value });
  }

  render() {
    const { id, name, required, placeholder, style } = this.props;
    const { street, city, neighborhood, uf, fullAddress } = this.state;

    return (
      <Fragment>
        <a href={'http://www.buscacep.correios.com.br'} target={'_blank'} className={'form__label-link'}  rel={'noopener noreferrer'}>NÃ£o lembra seu CEP?</a>
        <input id={id} name={name} className={style} type={'tel'} placeholder={placeholder} required={required} onKeyUp={this.onKeyUp} ref={this.ref}/>
        <span className={'full-address'}>{fullAddress}</span>
        <input id={'street'} name={'street'} type={'hidden'} value={street}/>
        <input id={'neighborhood'} name={'neighborhood'} type={'hidden'} value={neighborhood}/>
        <input id={'city'} name={'city'} type={'hidden'} value={city}/>
        <input id={'uf'} name={'uf'} type={'hidden'} value={uf}/>
      </Fragment>
    );
  }
}

Zipcode.propTypes = propTypes;
Zipcode.defaultProps = defaultProps;
