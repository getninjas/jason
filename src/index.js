import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

export default class Jason {
  constructor(options) {
    Object.assign(this, options);

    this.jason = null;
  }

  post() {
    this.jason.post();
  }

  init() {
    ReactDOM.render(
      <App data={this.data.form} {...this} ref={(jason) => { this.jason = jason; }} />,
      this.element,
    );
  }
}
