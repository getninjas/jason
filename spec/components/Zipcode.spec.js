import React from 'react';
import renderer from 'react-test-renderer';
import { enzymeConfig, mount } from '../enzymeConfig';
import { AppContext } from '../../src/AppContextReact';
import Zipcode from '../../src/components/Zipcode';

enzymeConfig();

const getLanguageSelectorWithContext = (context = { onZipcodeFetchSuccess: zipcode => zipcode }) => {
  jest.doMock('../../src/AppContextReact', () => {
    return {
      AppContext: {
        Consumer: (props) => {
          return (props.children(context))
        }
      }
    }
  });

  return require('../../src/components/Zipcode').default;
};

function createNodeMock(element) {
  if (element.type === 'input') {
    return {
      addEventListener() {},
      value: '',
    };
  }

  return null;
}

const commonProps = {
  type:'zipcode',
  key:'zipcode-1',
  id:'zipcodeTest',
  name:'zipcodeTest',
  placeholder:'00000-000',
  zipcodeUrlService:'http://viacep.com.br/ws/@@zipcode@@/json/',
  onFieldChange:() => {},
}

const zipcodeElement = () => {
  return <AppContext.Provider value={{ onZipcodeFetchSuccess: zipcode => zipcode }} >
    <Zipcode {...commonProps} />
  </AppContext.Provider>
};

const getZipCodeMock = () => {
  const ZipcodeMock = getLanguageSelectorWithContext();
  return (<ZipcodeMock {...commonProps}/>);
}

describe('Zipcode', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('renders defaultProps', () => {
    const options = { createNodeMock };
    const component = renderer.create(zipcodeElement(), options)

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('.getFullAddress', () => {
    it('returns formatted full address', () => {
      const component = mount(getZipCodeMock());

      const responseAddress = {
        street: 'Avenida Rebouças',
        neighborhood: 'Pinheiros',
        city: 'São Paulo',
        uf: 'SP',
      }

      const result = component.instance().getFullAddress(responseAddress);

      const fullAddressFormatted = 'Avenida Rebouças, Pinheiros \nSão Paulo - SP';

      expect(result).toEqual(fullAddressFormatted);
    });
  });

  describe('.onKeyUp', () => {
    it('calls valid zipcode methods', () => {
      const component = mount(getZipCodeMock());

      component.instance().isUserTyping = jest.fn();
      component.instance().isValidZipCodeInput = jest.fn();

      const evt = { target: { value: '05402-300' }, key: 0 }
      component.instance().onKeyUp(() => {}, evt);

      expect(component.instance().isUserTyping).toHaveBeenCalled();
      expect(component.instance().isValidZipCodeInput).toHaveBeenCalled();
    });

    it('starts zipcode fetch when fetchCompleted false', () => {
      const component = mount(getZipCodeMock());

      component.instance().getZipCode = jest.fn();

      const evt = { target: { value: '04707-060' }, key: 0 };
      const successCallback = () => {};
      component.instance().onKeyUp(successCallback, evt);

      expect(component.instance().getZipCode).toHaveBeenCalledWith('04707060', successCallback);
    });

    it('does not fetch zipcode if fetchCompleted true', () => {
      const component = mount(getZipCodeMock());

      component.state().fetchCompleted = true;
      component.instance().getZipCode = jest.fn();

      const evt = { target: { value: '04707-060' }, key: 0 };
      const successCallback = () => {};
      component.instance().onKeyUp(successCallback, evt);

      expect(component.instance().getZipCode).not.toHaveBeenCalledWith('04707060', successCallback);
    });

    it('calls onZipcodeSuccess and successCallback on fetch success', async () => {
      const component = mount(getZipCodeMock());

      const responseData = {
        data: { type_street: '', street: 'Rua Mock', city: 'Cidade Mock', neighborhood: 'Bairro Mock', uf: 'SP' }
      };
      const successCallback = jest.fn();
      component.instance().onZipcodeSuccess = jest.fn();

      await component.instance().getZipCode('04707060', successCallback);

      expect(successCallback).toHaveBeenCalledWith('04707060');
      expect(component.instance().onZipcodeSuccess).toHaveBeenCalledWith('04707060', responseData);
    });

    it('calls onZipcodeError on fetch error', async () => {
      const component = mount(getZipCodeMock());
      const successCallback = jest.fn();
      component.instance().onZipcodeError = jest.fn();
      component.state().zipcodeUrlService = 'http://unknowervice/@@zipcode@@';

      await component.instance().getZipCode('04707060', (zipcode) => { zipcode });

      expect(successCallback).not.toHaveBeenCalledWith('04707060');
      expect(component.instance().onZipcodeError).toHaveBeenCalledWith('04707060');
    });
  });

  describe('.isUserTyping', () => {
    it('returns true for incomplete zipcode format', () => {
      const zipcodeLength = 7;
      const keyboardKey = 'a';
      const component = mount(getZipCodeMock());

      const result = component.instance().isUserTyping(zipcodeLength, keyboardKey);

      expect(result).toEqual(true);
    });

    it('returns false when last zipcode caracter is entered', () => {
      const zipcodeLength = 8;
      const keyboardKey = 0;
      const component = mount(getZipCodeMock());

      const result = component.instance().isUserTyping(zipcodeLength, keyboardKey);

      expect(result).toEqual(false);
    });
  });

  describe('.isValidZipCodeInput', () => {
    it('returns true for a valid zipcode user input', () => {
      const zipcodeLength = 8;
      const fetchedCompleted = false;
      const component = mount(getZipCodeMock());

      const result = component.instance().isValidZipCodeInput(zipcodeLength, fetchedCompleted);

      expect(result).toEqual(true);
    });

    it('returns false for invalid zipcode user input', () => {
      const zipcodeLength = 7;
      const fetchedCompleted = true;
      const component = mount(getZipCodeMock());

      const result = component.instance().isValidZipCodeInput(zipcodeLength, fetchedCompleted);

      expect(result).toEqual(false);
    });
  });

  describe('.getEmptyState', () => {
    it('sets all key values to empty string', () => {
      const component = mount(getZipCodeMock());

      const currentState = {
        value: '05402300',
        type_street: '',
        street: 'Aveninda Rebouças',
        city: 'São Paulo',
        neighborhood: 'Pinheiros',
        uf: 'SP',
        fullAddress: 'Avenida Rebouças, Pinheiros \nSão Paulo - SP',
        fetchCompleted: true,
      }

      const emptyState = {
        value: '',
        type_street: '',
        street: '',
        city: '',
        neighborhood: '',
        uf: '',
        fullAddress: '',
        fetchCompleted: false,
      }

      const result = component.instance().getEmptyState(currentState);

      expect(result).toEqual(emptyState);
    });
  });

  describe('.fillAddressState', () => {
    it('sets response data and value zipcode to key values', () => {
      const component = mount(getZipCodeMock());
      const zipcodeValue = '05402300';

      const responseAddress = {
        type_street: '',
        street: 'Avenida Rebouças',
        city: 'São Paulo',
        neighborhood: 'Pinheiros',
        uf: 'SP',
      }

      const filledState = {
        value: '05402300',
        ...responseAddress,
        fullAddress: 'Avenida Rebouças, Pinheiros \nSão Paulo - SP',
        fetchCompleted: true,
      }

      const result = component.instance().fillAddressState(responseAddress, zipcodeValue);

      expect(result).toEqual(filledState);
    });
  });
});
