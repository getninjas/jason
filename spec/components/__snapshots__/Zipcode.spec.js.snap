// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zipcode renders defaultProps 1`] = `
Array [
  <a
    className="form__label-link"
    href="http://www.buscacep.correios.com.br"
    rel="noopener noreferrer"
    target="_blank"
>
    Não lembra seu CEP?
</a>,
  <input
    className="form__input"
    id="zipcodeTest"
    name="zipcodeTest"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="00000-000"
    required={true}
    type="tel"
    value=""
/>,
  <span
    className="full-address"
>
    
</span>,
  <input
    id="street"
    name="street"
    type="hidden"
    value=""
/>,
  <input
    id="neighborhood"
    name="neighborhood"
    type="hidden"
    value=""
/>,
  <input
    id="city"
    name="city"
    type="hidden"
    value=""
/>,
  <input
    id="uf"
    name="uf"
    type="hidden"
    value=""
/>,
]
`;

exports[`Zipcode triggers native DOM event 1`] = `
Array [
  <a
    className="form__label-link"
    href="http://www.buscacep.correios.com.br"
    rel="noopener noreferrer"
    target="_blank"
>
    Não lembra seu CEP?
</a>,
  <input
    className="form__input"
    id="zipcodeTest"
    name="zipcodeTest"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="00000-000"
    required={true}
    type="tel"
    value="04904-160"
/>,
  <span
    className="full-address"
>
    
</span>,
  <input
    id="street"
    name="street"
    type="hidden"
    value=""
/>,
  <input
    id="neighborhood"
    name="neighborhood"
    type="hidden"
    value=""
/>,
  <input
    id="city"
    name="city"
    type="hidden"
    value=""
/>,
  <input
    id="uf"
    name="uf"
    type="hidden"
    value=""
/>,
]
`;
